app/models/grda_warehouse/ch_enrollment.rb

```diff
@@ -74,5 +74,324 @@ def self.update_existing!
+      value = enrollment.TimesHomelessPastThreeYears
+
+      result = if THREE_OR_FEWER_TIMES_HOMELESS.include?(value)
+        :no
+      elsif dk_or_r_or_missing(value)
+        dk_or_r_or_missing(value)
+      end
+
+      { result: result, display_value: value }
+    end
+
+    # Lines 6, 12, 19, and 26 (3.917.4)
+    def self.total_months_homeless(enrollment, date: enrollment.EntryDate)
+      value = enrollment.MonthsHomelessPastThreeYears
+      return { result: :yes, display_value: value - 100 } if TWELVE_OR_MORE_MONTHS_HOMELESS.include?(value)
+
+      # If you don't have time prior to entry, day calculation above will catch any days during the enrollment
+      # If you have time prior to entry and we are looking at an arbitrary date, we need to add
+      # the months served. (This is only used for Chronic-at-PIT calculation, not Chronic-at-Entry).
+      if date != enrollment.EntryDate && enrollment.MonthsHomelessPastThreeYears.present? && enrollment.MonthsHomelessPastThreeYears > 100
+        project = enrollment.project
+        months_in_enrollment = if project.so? || project.es? && project.bed_night_tracking?
+          dates_in_enrollment_between(enrollment, enrollment.EntryDate, date).map do |d|
+            [d.month, d.year]
+          end.uniq.count
+        else
+          month_count = (date.year * 12 + date.month) - (enrollment.EntryDate.year * 12 + enrollment.EntryDate.month)
+          # Subtract 1 from this number if the [project start date] does not fall on the first of the month.
+          month_count -= 1 if month_count.positive? && enrollment.EntryDate.day != 1
+          month_count
+        end
+        months_prior_to_enrollment = enrollment.MonthsHomelessPastThreeYears - 100
+        sum = months_prior_to_enrollment + months_in_enrollment
+        return { result: :yes, display_value: sum } if sum > 11
+      end
+
+      return { result: dk_or_r_or_missing(value), display_value: value } if dk_or_r_or_missing(value)
+
+      { result: nil, display_value: value - 100 }
+    end
+
+    # TODO: test boundaries days/months for entry/exit, NbN, and SO
+    def self.homeless_duration_sufficient(enrollment, date: enrollment.EntryDate)
+      steps = [approximate_start_date(enrollment, date: date)]
+      return steps if steps.last[:result]
+
+      steps.push(num_times_homeless(enrollment))
+      return steps if steps.last[:result]
+
+      steps.push(total_months_homeless(enrollment, date: date))
+      steps
+    end
+
+    # Add steps for lines 15-16 and lines 22-23
+    def self.length_of_stay_previous_sufficient(enrollment)
+      steps = []
+      steps.push({ result: is_no?(enrollment.LOSUnderThreshold), display_value: enrollment.LOSUnderThreshold })
+      return steps if steps.last[:result]
+
+      steps.push({ result: is_no?(enrollment.PreviousStreetESSH), display_value: enrollment.PreviousStreetESSH })
+      steps
+    end
+
+    def self.dates_in_enrollment_between(enrollment, start_date, end_date)
```

Comment:
Since we call this twice, it's probably worth memoizing.
https://github.com/matthewrudy/memoist
```
class Person

  class << self
    extend Memoist
    def with_overdue_taxes
      # ...
    end
    memoize :with_overdue_taxes
  end

end
```

This can be a somewhat expensive DB fetch for larger histories.